photosynthesis_metrics/msid.py

```diff
@@ -0,0 +1,420 @@
+        normalized_laplacian: if True, use normalized Laplacian
+        normalize: 'empty' for average heat kernel (corresponds to the empty graph normalization of NetLSD),
+                'complete' for the complete, 'er' for erdos-renyi
+                normalization, 'none' for no normalization
+        msid_reduction: 'l2' to compute the l2 norm of the distance between `predictoon` and `target`;
+                'max' to find the maximum abosulute difference between two descriptors over temperature
+        channel_reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed. 
+
+    """
+
+    def __init__(self, ts: List[float] = np.logspace(-1, 1, 256), k: int = 5, m: int = 10, niters: int = 100, rademacher: bool = False, graph_builder: str = 'sparse',
+              normalized_laplacian: bool = True, normalize: str = 'empty', msid_reduction: str = "max", channel_reduction: str = 'mean'):
+
+        super(MSID, self).__init__()
+
+        self.msid_score = partial(msid_score, ts=ts, 
+                                                    k=k, 
+                                                    m=m, 
+                                                    niters=niters, 
+                                                    rademacher=rademacher, 
+                                                    graph_builder=graph_builder, 
+                                                    msid_mode=msid_reduction,
+                                                    normalized_laplacian=normalized_laplacian, 
+                                                    normalize=normalize)
+
+
+        # Generic loss parameters.
+        self.channel_reduction = channel_reduction
+
+        # Loss-specific parameters.
+
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor = None) -> torch.Tensor:
+        r"""Computation of MSID
+
+        Args:
+            x: Batch of images. Required to be 4D, channels first (N,C,H,W).
+            y: Batch of images. Required to be 4D, channels first (N,C,H,W).
+
+        Returns:
+            normed_msidx: normalized msid descriptor (if no `target` specified).
+            msid_score: the scalar value of the distance between discriptor if both `prediction` and `target` given.
+        """
+
+        N, _, _, _ = x.shape
+        result = []
+        for i in range(N):
+            x_i = x[i]
+            y_i = y[i]
+            
+            channels_x = [t.numpy() for t in map(torch.Tensor.squeeze_, torch.split(x_i, 1, dim=0))]
+            channels_y = [t.numpy() for t in map(torch.Tensor.squeeze_, torch.split(y_i, 1, dim=0))]
+
+            scores = torch.Tensor([self.msid_score(x, y) for x, y in zip(channels_x, channels_y)])
```

Comment:
Please read MSID paper carefully and find that graphs should be built on embeddings